Add a performance breakdown page for add-to-app

xster opened the pull request:
Also moved the files around a bit to allow room for cross-platform topical pages. 

Preview at https://add-to-app-2019.firebaseapp.com/docs/development/add-to-app/performance

Fixes https://github.com/flutter/flutter/issues/33002
Fixes https://github.com/flutter/flutter/issues/32949

This whole thing is mostly a simplification of https://docs.google.com/document/d/1Ql792KQ6SinpRYkQqxSikZmLprgJFPpxvVwsJD--C9o

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+
+## Loading Flutter
+
+Both Android/iOS (the 2 supported platforms for integrating into existing
```
You say "Both", but then list 3 options, unless I am misunderstanding. How about:

"Android and iOS apps (the two supported platforms for integrating into existing apps),
full Flutter apps, and add-to-app patterns have a similar sequence of conceptual 
loading steps when displaying Flutter UI."

xster replied:
I meant both dimensions / duos. But your version is better.

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+On **Android**, this happens also when the [`FlutterEngine`]({{site.api}}/javadoc/io/flutter/embedding/engine/FlutterEngine.html)
+is constructed since the JNI connectors need to reference the Flutter C++
+library. On **iOS**, this happens when the [`FlutterEngine`]({{site.api}}/objcdoc/Classes/FlutterEngine.html)
+is first run, such as via [`runWithEntrypoint:`]({{site.api}}/objcdoc/Classes/FlutterEngine.html#/c:objc(cs)FlutterEngine(im)runWithEntrypoint:).
```
We don't use "via".  https://developers.google.com/style/word-list#letter-V

such as via => for example, when calling

xster replied:
Ah, I keep forgetting about that doc

xster replied:
Changed here and elsewhere (in the next commit)

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+
+### Starting the Dart VM
+
+The Dart runtime is responsible for managing Dart memory, concurrency etc. for
```
Can you delete "etc."? It's a bit jarring. Then:

memory, concurrency, etc. => memory and concurrency

xster replied:
ok

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+
+This is done by starting a Dart Isolate in the Dart runtime. The Isolate is
+Dart's container for memory and thread. A number of [auxiliary threads](https://github.com/flutter/flutter/wiki/The-Engine-architecture#threading)
+on the host platform is also created at this point to support the Dart Isolate,
```
is also created => are also created

to support the Dart Isolate => to support the isolate,

xster replied:
ok

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+This is done by starting a Dart Isolate in the Dart runtime. The Isolate is
+Dart's container for memory and thread. A number of [auxiliary threads](https://github.com/flutter/flutter/wiki/The-Engine-architecture#threading)
+on the host platform is also created at this point to support the Dart Isolate,
+such as offloading GPU handling and image decoding off of the Dart Isolate.
```
=> such as a thread for offloading GPU handling and another for image decoding.

xster replied:
ok

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+on the host platform is also created at this point to support the Dart Isolate,
+such as offloading GPU handling and image decoding off of the Dart Isolate.
+
+One Dart Isolate exists per `FlutterEngine` instance and multiple Dart Isolates
```
I don't understand why it's always "Dart Isolate", which seems very formal and repetitive. Why not just "isolate"? (Once it's been introduced.)

xster replied:
sure

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+At this point, your Dart code's selected entrypoint (the `main()` function of
+your Dart library's main.dart file by default) is executed. If you called the
+Flutter function [runApp()]({{site.api}}/flutter/widgets/runApp.html) in your
+`main()` function, your Flutter app or library's widget tree will also be
```
We avoid future tense, if possible.  https://developers.google.com/style/tense

"widget tree will also be" => "widget tree are also"

xster replied:
done

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+Flutter function [runApp()]({{site.api}}/flutter/widgets/runApp.html) in your
+`main()` function, your Flutter app or library's widget tree will also be
+created and built. If you need to gate certain functionalities from executing
+in your Flutter code, the `AppLifecycleState.detached` enum value will indicate
```
I think it would be more clear if you replaced "gate" with "prevent".

"enum value will indicate" => "enum value indicates"

xster replied:
done

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+In an add-to-app scenario, this happens when you attach a `FlutterEngine`
+to a UI component such as by calling [`startActivity()`](https://developer.android.com/reference/android/content/Context.html#startActivity(android.content.Intent))
+with an [`Intent`](https://developer.android.com/reference/android/content/Intent.html)
+built via [`FlutterActivity.withCachedEngine()`]({{site.api}}/javadoc/io/flutter/embedding/android/FlutterActivity.html#withCachedEngine-java.lang.String-)
```
build via => built with

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+with an [`Intent`](https://developer.android.com/reference/android/content/Intent.html)
+built via [`FlutterActivity.withCachedEngine()`]({{site.api}}/javadoc/io/flutter/embedding/android/FlutterActivity.html#withCachedEngine-java.lang.String-)
+on **Android**. Or by presenting a [`FlutterViewController`]({{site.api}}/objcdoc/Classes/FlutterViewController.html)
+initialized via [`initWithEngine: nibName: bundle:`]({{site.api}}/objcdoc/Classes/FlutterViewController.html#/c:objc(cs)FlutterViewController(im)initWithEngine:nibName:bundle:)
```
initialized via => initialized by using

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+on **iOS**.
+
+This would also the case if a Flutter UI component was launched without
+pre-warming a `FlutterEngine` such as via [`FlutterActivity.createDefaultIntent()`]({{site.api}}/javadoc/io/flutter/embedding/android/FlutterActivity.html#createDefaultIntent-android.content.Context-)
```
such as via => such as by using

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+This would also the case if a Flutter UI component was launched without
+pre-warming a `FlutterEngine` such as via [`FlutterActivity.createDefaultIntent()`]({{site.api}}/javadoc/io/flutter/embedding/android/FlutterActivity.html#createDefaultIntent-android.content.Context-)
+on **Android** or via [`FlutterViewController initWithProject: nibName: bundle:`]({{site.api}}/objcdoc/Classes/FlutterViewController.html#/c:objc(cs)FlutterViewController(im)initWithProject:nibName:bundle:)
+on **iOS**. An implicit `FlutterEngine` is created in these cases. ;
```
Delete the semi-colon at the end.

xster replied:
oops, sorry

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+on **Android** or via [`FlutterViewController initWithProject: nibName: bundle:`]({{site.api}}/objcdoc/Classes/FlutterViewController.html#/c:objc(cs)FlutterViewController(im)initWithProject:nibName:bundle:)
+on **iOS**. An implicit `FlutterEngine` is created in these cases. ;
+
+Behind the scene, both platform's UI components serves to provide the
```
"both platform's UI components" ?? What two components are you talking about?

Also, if it really is two platform components, it would be "serve", not "serves".

xster replied:
I meant the Android and iOS platforms' UI components

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+on **iOS**.
+
+At this point, the [Layer]({{site.api}}/flutter/rendering/Layer-class.html)
+tree generated by your Flutter program per frame will be converted into
```
"your Flutter program per frame" ??  Is there a layer tree generated for each frame?

"will be" => "is"

xster replied:
There is. Done

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+
+## Memory and latency
+
+Showing a Flutter UI has non-trivial latency cost. This cost can be alleviated
```
Just curious. Why do you prefer "showing a Flutter UI" to "displaying a Flutter UI"? Is there some meaningful difference? I just find "displaying" to be clearer.

xster replied:
This may be completely unjustified. There's a little tingle in the back of my head that makes me think that displaying is a slightly smaller subset of what it takes to show it. As in, to show you results, I'd have to produce it first whereas displaying is slightly more confined to the broadcasting component of it. 

In this case, the non-trivial cost is really the entire process leading up to Flutter rendering, not just the rendering part of it.

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+
+The most relevant choice for users in add-to-app scenarios is to decide
+when to pre-load a `FlutterEngine` (that is, to load the Flutter library,
+start the Dart VM and run entrypoint in an Isolate) and the memory and latency
```
start the Dart VM and run entrypoint in an Isolate => start the Dart VM, and run the entrypoint in an isolate,

xster replied:
done

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+The most relevant choice for users in add-to-app scenarios is to decide
+when to pre-load a `FlutterEngine` (that is, to load the Flutter library,
+start the Dart VM and run entrypoint in an Isolate) and the memory and latency
+cost of that pre-warm. And how it affects the memory and latency cost of
```
And how it... => Also, how it affects the memory and latency cost

xster replied:
done

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+when to pre-load a `FlutterEngine` (that is, to load the Flutter library,
+start the Dart VM and run entrypoint in an Isolate) and the memory and latency
+cost of that pre-warm. And how it affects the memory and latency cost of
+rendering a first Flutter frame when UI is subsequently attached to that
```
rendering a first Flutter frame when UI => rendering a first Flutter frame when **the** UI

xster replied:
changed to `the UI component`

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+rendering a first Flutter frame when UI is subsequently attached to that
+`FlutterEngine`.
+
+As of September of 2019, testing on a low-end 2015 class device in release-AOT
```
September of 2019 => September 2019

Could we say "As of Flutter 1.9..." ? Or is this referring to the master channel for 1.10.x?

xster replied:
Sure. That's probably better. Changed to `As of Flutter v1.10.3`

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+As of September of 2019, testing on a low-end 2015 class device in release-AOT
+mode, pre-warming the `FlutterEngine` costs:
+
+- 42MB and 1530ms to prewarm on **Android**. 330ms of it is a blocking call on
```
330ms of it is a blocking => 330ms is due to a blocking

xster replied:
Rather than trying to explain the previous number's composition, I'm just expanding on another subcomponent which is of interest in its own right

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+
+- 42MB and 1530ms to prewarm on **Android**. 330ms of it is a blocking call on
+the main thread.
+- 22MB and 860ms to prewarm on **iOS**. 260ms of it is a blocking call on the
```
260ms ofit is a blocking => 260ms is due to a blocking

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+Flutter engine start-up time into the first frame latency of showing Flutter.
+
+The exact timing depends on the app's structure and heuristic. An example would
+be to load the Flutter engine in the screen before the screen drawn by Flutter.
```
before the screen drawn => before Flutter draws the screen

xster replied:
the thing that's before is a position (the other screen), rather than a time (when Flutter draws)

sfshaza2 commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+that are not pre-emptively predictable. This is a one-time cost.
+
+When the Flutter UI component is released, the UI related memory is freed.
+This doesn't affect the Flutter state which lives in the `FlutterEngine`
```
state which => state, which

xster replied:
done

xster commented:
great, thanks so much for reviewing. Will address

matthew-carroll commented on src/_data/sidenav.yml:
```diff
@@ -125,18 +125,22 @@
-            - title: Add a Splash Screen
-              permalink: /docs/development/platform-integration/add-to-app-android/add-splash-screen
+          - title: Adding to an Android app
+            permalink: /docs/development/platform-integration/add-to-app/android
```
This link seems to be broken in the staging website. Also, due to how low this has been moved, the individual Android options can no longer be seen in the side bar, which I think we probably should revert. I know that I'd certainly want to see the individual topics in the navigation area.

xster replied:
ya, I noticed last night too. 

@Sfshaza2 should I try to tweak this in the implementation or is the depth intentionally fixed to this level?

sfshaza2 commented:
The depth was intentionally set to that level, but I believe it's a fairly arbitrary value.. I would be fine with adding a level, if you want to attempt it.

xster commented on src/docs/development/platform-integration/add-to-app/performance.md:
```diff
@@ -0,0 +1,174 @@
+title: Load sequence, performance and memory
+---
+
+This page serves to inform the reader of the breakdown of the steps involved
```
did you mean referring to ourselves in the third person (I think that's not what you meant since your suggestion does it too) or referring to the reader in the third person?

xster commented:
I tried mucking with it a bit. Unfortunately the code controlling the side nav depth is spread across 10 files. So I moved the files around a bit within the 3 levels and moved add-to-app up a layer. 

I suspect this should be ok since there'll be 10+ pages in there by the time we're done. It might be a bit cramped if it were all stuffed into the lowest layer of the nav tree. WDYT?

sfshaza2 commented:
That sounds reasonable to me, Xiao. I will definitely take a look at the
sidenav as things progress.

On Wed, Sep 18, 2019 at 1:42 AM xster <notifications@github.com> wrote:

> I tried mucking with it a bit. Unfortunately the code controlling the side
> nav depth is spread across 10 files. So I moved the files around a bit
> within the 3 levels and moved add-to-app up a layer.
>
> I suspect this should be ok since there'll be 10+ pages in there by the
> time we're done. It might be a bit cramped if it were all stuffed into the
> lowest layer of the nav tree. WDYT?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/website/pull/3007?email_source=notifications&email_token=AKS4PKPYX2I7AW2BAHBP7OTQKHSWTA5CNFSM4IXOYELKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD67JRQQ#issuecomment-532584642>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKS4PKM2MCIAW2NE4TTF3TTQKHSWTANCNFSM4IXOYELA>
> .
>

xster commented:
SG. I did that described move in this PR. You can check on https://add-to-app-2019.firebaseapp.com/docs/development/add-to-app/performance
